service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/bucket/CopyGcsBucketDefinitionToReferencedStep.java

```diff
@@ -0,0 +1,134 @@
+ * resource. Bucket namd and creation parameters (lifecycle, storage class, etc) are copied from the
+ * source bucket. CLONED_RESOURCE_DEFINITION is put into the working map for undo(). 'The response
+ * is set on the flight.
+ *
+ * <p>Keep in sync with CopyGcsBucketDefinitionToControlledStep.
+ */
+public class CopyGcsBucketDefinitionToReferencedStep implements Step {
+
+  private final AuthenticatedUserRequest userRequest;
+  private final ControlledGcsBucketResource sourceBucket;
+  private final ReferencedResourceService referencedResourceService;
+  private final CloningInstructions resolvedCloningInstructions;
+
+  public CopyGcsBucketDefinitionToReferencedStep(
+      AuthenticatedUserRequest userRequest,
+      ControlledGcsBucketResource sourceBucket,
+      ReferencedResourceService referencedResourceService,
+      CloningInstructions resolvedCloningInstructions) {
+    this.userRequest = userRequest;
+    this.sourceBucket = sourceBucket;
+    this.referencedResourceService = referencedResourceService;
+    this.resolvedCloningInstructions = resolvedCloningInstructions;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+    final FlightMap inputParameters = flightContext.getInputParameters();
+    final FlightMap workingMap = flightContext.getWorkingMap();
+    FlightUtils.validateRequiredEntries(
+        inputParameters,
+        ControlledResourceKeys.DESTINATION_WORKSPACE_ID,
+        ControlledResourceKeys.DESTINATION_RESOURCE_ID);
+    Preconditions.checkState(
+        resolvedCloningInstructions == CloningInstructions.COPY_REFERENCE,
+        "CloningInstructions must be COPY_REFERENCE");
+    final String resourceName =
+        FlightUtils.getInputParameterOrWorkingValue(
+            flightContext,
+            ResourceKeys.RESOURCE_NAME,
+            ResourceKeys.PREVIOUS_RESOURCE_NAME,
+            String.class);
+    final String description =
+        FlightUtils.getInputParameterOrWorkingValue(
+            flightContext,
+            ResourceKeys.RESOURCE_DESCRIPTION,
+            ResourceKeys.PREVIOUS_RESOURCE_DESCRIPTION,
+            String.class);
+    final UUID destinationWorkspaceId =
+        inputParameters.get(ControlledResourceKeys.DESTINATION_WORKSPACE_ID, UUID.class);
+    final var destinationResourceId =
+        inputParameters.get(ControlledResourceKeys.DESTINATION_RESOURCE_ID, UUID.class);
+
+    ReferencedGcsBucketResource destinationBucketResource =
+        WorkspaceCloneUtils.buildDestinationReferencedGcsBucketFromControlled(
+            sourceBucket,
+            destinationWorkspaceId,
+            destinationResourceId,
+            resourceName,
+            description,
+            sourceBucket.getBucketName());
+
+    // Launch a CreateReferenceResourceFlight to make the destination bucket
+    final ReferencedGcsBucketResource createdBucket =
```

Comment:
This was not the right approach when Jay did it and it is still not the right approach. You cannot easily use flights as subroutines.

In this case, a failure-rerun of this step will launch a new create flight that is likely to fail because the referenced resource will already exist.

I probably shouldn't have let him off the hook with an "I'll fix it later". Let's not compound the problem.

I think the better approach would be to make shared code that generates the steps for creating a referenced resource. Use that step generator in this flight and do the referenced resource creation in line. Also use it in the main createReferencedResource method.

That is, rather than trying to share the flight, share the code generation for the steps.

Reply:
Done.

Old: Flight had CopyGcsBucketDefinitionStep, which started a sub-flight

New: Flight has SetReferencedDestinationGcsBucketWorkingMapStep, CreateReferenceMetadataStep, SetReferencedDestinationGcsBucketResponseStep. No sub-flight

Added tests to ControlledResourceServiceTest:

- cloneGcsBucket_copyResource_do
- cloneGcsBucket_copyReference_do
- cloneGcsBucket_copyResource_undo
- cloneGcsBucket_copyReference_undo

For the do tests, I set above three steps to STEP_RESULT_FAILURE_RETRY.

(The undo tests were in CloneControlledGcsBucketResourceFlightTest; I moved them to ControlledResourceServiceTest and deleted CloneControlledGcsBucketResourceFlightTest.)